subsys/bluetooth/shell/has.c

```diff
@@ -0,0 +1,102 @@
+/**
+ * @file
+ * @brief Bluetooth Hearing Access Service (HAS) shell.
+ *
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/audio/has.h>
+#include <shell/shell.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "bt.h"
+
+static int cmd_preset_reg(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+	struct bt_has_preset_register_param param = {
+		.index = strtol(argv[1], NULL, 0),
+		.properties = strtol(argv[2], NULL, 0),
+		.name = argv[3],
+	};
+
+	err = bt_has_preset_register(&param);
+	if (err < 0) {
+		shell_error(sh, "Preset register failed (err %d)", err);
+		return err;
+	}
+
+	return 0;
+}
+
+static int cmd_preset_unreg(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+	const int index = strtol(argv[1], NULL, 0);
+
+	err = bt_has_preset_unregister(index);
+	if (err < 0) {
+		shell_print(sh, "Preset unregister failed (err %d)", err);
+		return err;
+	}
+
+	return 0;
+}
+
+struct print_list_entry_data {
+	int num;
+	const struct shell *sh;
+};
```

Comment:
That's also a way of handling the shell in the callback :) 

For most part we just assign `ctx_shell` and use that in the callbacks, but either way works here. I just hope that this wasn't the only reason why `bt_has_preset_foreach` takes a `user_data` parameter :D

Reply:
Nope, haha :D